FIX-#4527: Log Modin Ray execution

jeffreykennethli opened the pull request:
<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/development/contributing.html
if you have questions about contributing.
-->

## What do these changes do?

Adds logging (via `ClassLogger`) to the Partition layer (partition manager, partition, axis partition layer), and execution (Ray remote functions), and disables logging for callback-functions like `read_callback` and `deploy_axis_func`, which are called from within a Ray worker and should be logged there.

Note:
Due to the complex interaction between python logging, Ray workers, and Modin, it is necessary to set the runtime environment variable for the Ray tasks. Ray tasks execute as separate processes, and while it is possible to share data between processes, such as a logger instance, file handler, and env vars like whether `LogMode == enabled`, it would require a rewrite of our interface with Ray. Currently, this implementation will set `LogMode` once for each task but never update it. This seems bad, but is also basically the behavior or Ray timeline (assumes envvars are static).

To run/test:
Run a script with `LogMode.enable()` that spawns Ray workers, such as `read_csv` on a semi-large file. I would also recommend limiting `NPartitions` to be 4 to reduce the number of spawned workers for testing. The resultant log file structure should be:

```
.modin/logs/job_<uuid>
    trace.log
    errors.log
    memory.log
    <worker_id_1>.log
    <worker_id_2>.log
```

<!-- Please give a short brief about these changes. -->

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [ ] Resolves #4527 <!-- issue must be created for each patch -->
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->
- [ ] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release <!-- e.g. DOCS-#4077: Add release notes template to docs folder (#4078) -->

vnlitvinov commented on modin/core/dataframe/pandas/partitioning/axis_partition.py:
```diff
@@ -125,6 +126,7 @@ def shuffle(self, func, lengths, **kwargs):
         return self._wrap_partitions(self.deploy_axis_func(*args, **kwargs))
 
     @classmethod
+    @disable_logging
```
can you add a comment why it's disabled?

vnlitvinov commented on modin/core/execution/ray/common/task_wrapper.py:
```diff
@@ -18,11 +18,15 @@
+from modin.logging.config import JOB_ID
 import ray
 
+__WORKER_ENV__ = {"env_vars": {"MODIN_LOG_MODE": LogMode.get()}}
```
please add a note to the docs about "set before running, cannot change later" behaviour

vnlitvinov commented on modin/core/execution/ray/common/task_wrapper.py:
```diff
@@ -39,10 +43,14 @@ def _deploy_ray_func(func, args):  # pragma: no cover
-    return func(**args)
+    logger = get_worker_logger(JOB_ID)
+    logger.info(f"START::PANDAS-API::deploy_ray_func.{func.__qualname__}")
+    res = func(**args)
```
I see no exception handling, it would lead to logging unpaired `START` if any error is raised in `func`

vnlitvinov commented on modin/core/execution/ray/common/task_wrapper.py:
```diff
@@ -39,10 +43,14 @@ def _deploy_ray_func(func, args):  # pragma: no cover
     """
-    return func(**args)
+    logger = get_worker_logger(JOB_ID)
+    logger.info(f"START::PANDAS-API::deploy_ray_func.{func.__qualname__}")
```
```suggestion
    logger.info(f"START::RAY-REMOTE::deploy_ray_func:{func.__qualname__}")
```
I suggest changing both "modin layer" to signify it's Ray and to change dot to colon to highlight that `func.__qualname__` is not a child of `deploy_ray_func` (however the latter is questionable as it could potentially break log parsing if it exists already)

vnlitvinov commented on modin/core/dataframe/base/partitioning/axis_partition.py:
```diff
@@ -14,9 +14,10 @@
 
 
-class BaseDataframeAxisPartition(ABC):  # pragma: no cover
+class BaseDataframeAxisPartition(ClassLogger, ABC):  # pragma: no cover
```
I wonder if we should set some specific `modin_layer="PARTITION"` logging layer here (as `ClassLogger` supports this), to make following logs easier

Here and in other classes, too. Stuffing everything under `PANDAS-API` umbrella feels wrong.

devin-petersohn replied:
+1 to @vnlitvinov I think we want to make sure that log column has some meaning 😄

jeffreykennethli replied:
I tried doing `modin_layer="PARTITION"` in the class declaration as you mentioned, but the `modin_layer` variable was not being propagated to all subclasses. E.g. `PandasDataframePartition` would be `PARTITION` but `PandasOnRayDataframePartition` would be `PANDAS-API`.

I think the reason is that `__init_subclass__` is called from the perspective of `cls` (`PandasOnRayDataframePartition`) to `ClassLogger`, so it doesn't have the `modin_layer` of its immediate parent, `PandasDataframePartition`. Perhaps we could add this to the class declaration for each subclass. What I've implemented is another way, which is to "propagate" the var by making it a class variable, and making `ClassLogger` retrieve that class variable.

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:
```diff
@@ -54,16 +58,6 @@ def __init__(self, data, length=None, width=None, ip=None, call_queue=None):
         self._ip_cache = ip
         self._identity = uuid.uuid4().hex
 
-        logger = get_logger()
```
Is there a reason to drop debug logging? `ClassLogger` only logs calls/exits at `.info` and it doesn't log particular details which might be interesting. I would just change the logging from eager formatting to lazy one (specifying the message template and format args as separate things, like `logger.debug("My message: {height}", height=self._length_cache)`, so `logger` would compose the message only when debug is enabled.

Here and below.

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:
```diff
@@ -345,10 +323,18 @@ def _get_index_and_columns(df):
         The number of columns.
     """
-    return len(df.index), len(df.columns)
+    logger = get_worker_logger(JOB_ID)
+    logger.info(
+        f"START::PANDAS-API::PandasOnRayDataframePartition._get_index_and_columns"
+    )
+    res = len(df.index), len(df.columns)
+    logger.info(
+        f"STOP::PANDAS-API::PandasOnRayDataframePartition._get_index_and_columns"
+    )
+    return res
```
I would rather we implement specific worker logger decorator which would do this instead of copy-pasting this pattern all over the place... plus why `PANDAS-API`?..

jeffreykennethli replied:
Added a decorator `enable_remote_logging` which does this.

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:
```diff
@@ -375,22 +361,28 @@ def _apply_func(partition, func, *args, **kwargs):  # pragma: no cover
         get_node_ip_address(),
     )
+    logger.info(f"STOP::PANDAS-API::PandasOnRayDataframePartition._apply_func")
+    return res
```
you can keep `return` statement and add logging by using pattern like
```python
try:
    return foo
finally:
    logger.info("bar")
```
it's also better in terms of any exception raised during `return foo` line (in our case, for example, an exception in `get_node_ip_address()`) would _not_ prevent logging `STOP` message

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py:
```diff
@@ -504,7 +507,10 @@ class PandasOnRayDataframeRowPartition(PandasOnRayDataframeVirtualPartition):
 
 
-@ray.remote
+__WORKER_ENV__ = {"env_vars": {"MODIN_LOG_MODE": LogMode.get()}}
```
I would suggest to create this in `modin/core/execution/ray/common/utils.py` and import from that everywhere it's needed... or even create a custom decorator in there, like
```python
import ray
import functools

from modin.config import LogMode

__WORKER_ENV__ = {"env_vars": {"MODIN_LOG_MODE": LogMode.get()}}

logged_remote = functools.partial(ray.remote, runtime_env=__WORKER_ENV__)
```
and use it like
```python
from modin.core.execution.ray.common.utils import logged_remote

@logged_remote(num_returns=2)
def func(args):
    # do stuff
    return a, b
```

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py:
```diff
@@ -527,13 +533,21 @@ def deploy_ray_func(func, *args, **kwargs):  # pragma: no cover
     else:
-        return [i for r in result for i in [r, None, None, ip]]
+        res = [i for r in result for i in [r, None, None, ip]]
+    logger.info(
```
same about using
```python
try:
    if a: return foo
    if b: return bar
finally:
    logger.info("baz")
```
pattern

vnlitvinov commented on modin/core/storage_formats/pandas/parsers.py:
```diff
@@ -140,6 +141,7 @@ def find_common_type_cat(types):
 class PandasParser(ClassLogger):
     """Base class for parser classes with pandas storage format."""
 
+    @disable_logging
```
why?

jeffreykennethli replied:
I don't think this needs to be logged, will remove.

vnlitvinov commented on modin/pandas/accessor.py:
```diff
@@ -178,6 +179,7 @@ def to_dense(self):
 
 @_inherit_docstrings(pandas.core.accessor.CachedAccessor)
 class CachedAccessor(ClassLogger):
+    @disable_logging
```
why?

jeffreykennethli replied:
Removing this.

vnlitvinov reviewed (commented):
Overall looks good, left a few comments, though.

And please fix the linting etc in the CI.

jeffreykennethli commented on modin/logging/__init__.py:
```diff
@@ -12,12 +12,16 @@
-from .config import get_logger  # noqa: F401
-from .logger_decorator import enable_logging, disable_logging  # noqa: F401
+from .config import get_logger, get_worker_logger  # noqa: F401
+from .logger_decorator import (
+    enable_logging,
+    disable_logging,
+)  # noqa: F401
```
I think this was a black auto-format thing. I'll resolve it somehow!

devin-petersohn reviewed (commented):
I think this looks good. Will let @vnlitvinov be the final approval since he has more context.

jeffreykennethli commented:
Per @vnlitvinov this PR can be refactored so that the Ray runtime env is set when we call `ray.init` instead of at the Task level init. Marking this PR as a draft until this is done.

jeffreykennethli commented:
Rebased on master to get #4583 and #4620. Now the inheritance is done as originally designed (not through a class variable), and type checking is added to `logging`. I also moved the Ray task `runtime_env` variable to the `ray.init` kwargs, to avoid having to use a partial in place of all `@ray.remote` decorators.

vnlitvinov commented on docs/release_notes/release_notes-0.16.0.rst:
```diff
@@ -17,7 +17,7 @@ Key Features and Updates
   * PERF-#4182: Add cell-wise execution for binary ops, fix bin ops for empty dataframes (#4391)
   * PERF-#4288: Improve perf of `groupby.mean` for narrow data (#4591)
 * Benchmarking enhancements
-  *
+  * FIX-#4527: Log Modin Ray execution (#4587)
```
why it's under benchmarking?.. I think it's a feature

vnlitvinov commented on docs/usage_guide/advanced_usage/modin_logging.rst:
```diff
@@ -30,9 +30,10 @@ In the example below, we enable logging for internal Modin API calls.
+In the next example, we add logging for not only internal Modin API calls, but also for partition metadata, remote execution, and memory profiling.
 We can set the granularity (in seconds) at which the system memory utilization is logged using ``LogMemoryInterval``.
 We can also set the maximum size of the logs (in MBs) using ``LogFileSize``.
+**Note:**: Logging should be enabled before any Modin calls, and remote execution calls cannot be disabled once enabled.
```
```suggestion
**Note:**: Logging should be enabled before any Modin calls, and logging of remote execution calls cannot be disabled once enabled.
```

vnlitvinov commented on modin/core/execution/ray/common/utils.py:
```diff
@@ -206,7 +208,14 @@ def initialize_ray(
                 "object_store_memory": object_store_memory,
                 "_redis_password": redis_password,
                 "_memory": object_store_memory,
+                "runtime_env": {
+                    "env_vars": {
+                        "MODIN_LOG_MODE": LogMode.get(),
+                        "MODIN_LOG_FILE_SIZE": str(LogFileSize.get()),
+                    }
+                },
```
I wonder if we should first merge https://github.com/modin-project/modin/pull/4603 and then augment the passed dictionary there (if logging is enabled) so that the configuration is checked for the case of externally-started clusters?

vnlitvinov commented on modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:
```diff
@@ -53,15 +53,13 @@ def __init__(self, data, length=None, width=None, ip=None, call_queue=None):
-            )
+            "Partition ID: %s, Height: %s, Width: %s, Node IP: %s",
+            self._identity,
+            str(self._length_cache),
+            str(self._width_cache),
+            str(self._ip_cache),
```
```suggestion
            self._length_cache,
            self._width_cache,
            self._ip_cache,
```
no need to explicitly `str()` things, as `%s` formatter calls that for us

vnlitvinov commented on modin/core/storage_formats/pandas/parsers.py:
```diff
@@ -292,7 +293,9 @@ def single_worker_read(cls, fname, **kwargs):
 class PandasCSVParser(PandasParser):
+    # Disable logging for a function that is executed in a remote context, see TextFileDispatcher._launch_tasks
     @staticmethod
+    @disable_logging
```
this particular piece feels weird, could you please provide some more verbose reasoning behind it?

jeffreykennethli replied:
Added more context for this.

vnlitvinov replied:
I wonder what happens in non-remote context...

jeffreykennethli replied:
In a non-remote context (running a script that does call `PandasCSVParser.parse` without the remote logging enabled), the parse is not logged since it's invoked in a process (Ray process) that does not have access to the Modin logger.

vnlitvinov commented on modin/logging/logger_decorator.py:
```diff
@@ -143,3 +144,66 @@ def run_and_log(*args: Tuple, **kwargs: Dict) -> Any:
+            return func(*args, **kwargs)
+
+        logger = get_worker_logger(JOB_ID)
+        start_line = f"START::{modin_layer.upper()}::{func.__name__}"
+        stop_line = f"STOP::{modin_layer.upper()}::{func.__name__}"
```
these two should be computed outside of `run_and_log` (see `enable_logging` implementation), they never change, we shouldn't be wasting time computing same things over and over again

vnlitvinov replied:
this feels super weird, got a stacktrace?..

jeffreykennethli replied:
https://gist.github.com/jeffreykennethli/4cd7606e68718000b9467e9f47b6c9a0. also see https://github.com/jeffreykennethli/modin/blob/log_engine_2/modin/logging/logger_decorator.py#L149-L215

vnlitvinov commented on modin/logging/logger_decorator.py:
```diff
@@ -143,3 +144,66 @@ def run_and_log(*args: Tuple, **kwargs: Dict) -> Any:
+
+        funcs = [arg for arg in args if callable(arg)]
+        if funcs:
+            start_line += f":{funcs}"
+            stop_line += f":{funcs}"
```
```suggestion
            start_line += ":".join(f.__name__ for f in funcs)
            stop_line += ":".join(f.__name__ for f in funcs)
```
so we get colon-separated functions?.. or any other join, 'cause right now what we would end up with is something like `"START::RAY-REMOTE::my_func:'[<function foo1 at 0x000002B3A2B5F040>, <function foo2 at 0x000002B3F2B5F040>]'"` which looks weird (IMHO)

vnlitvinov commented on modin/core/execution/ray/common/task_wrapper.py:
```diff
@@ -18,11 +18,12 @@
 
 
-@ray.remote
+@ray.remote(num_returns=2)
```
why change?.. also, as this is a general wrapper, I don't think we can know upfront how many items it would return

vnlitvinov commented on docs/release_notes/release_notes-0.16.0.rst:
```diff
@@ -24,7 +24,6 @@ Key Features and Updates
   * PERF-#4288: Improve perf of `groupby.mean` for narrow data (#4591)
   * PERF-#4325: Improve perf of multi-column assignment in `__setitem__` when no new column names are assigning (#4455)
 * Benchmarking enhancements
-  *
```
why?

vnlitvinov commented on modin/core/execution/ray/common/task_wrapper.py:
```diff
@@ -18,11 +18,12 @@
 
 import asyncio
-
+from modin.logging import ClassLogger, enable_remote_logging
 import ray
```
we should import non-Modin first, then empty line, then Modin-defined modules

vnlitvinov commented on modin/core/execution/ray/common/utils.py:
```diff
@@ -41,6 +43,60 @@
     ObjectIDType = (ray.ObjectRef, ClientObjectRef)
 
 
+def _move_stdlib_ahead_of_site_packages(*args):
```
this was removed in #4603, pls not re-add it

vnlitvinov commented on modin/core/execution/ray/common/utils.py:
```diff
@@ -41,6 +43,60 @@
+        sys.path.insert(site_packages_path_index, os.path.dirname(site_packages_path))
+
+
+def _import_pandas(*args):
```
again, pls not re-add what was removed in #4603

vnlitvinov commented on modin/core/execution/ray/common/utils.py:
```diff
@@ -64,7 +120,15 @@ def initialize_ray(
+        "runtime_env": {
+            "env_vars": {
+                "__MODIN_AUTOIMPORT_PANDAS__": "1",
+                "MODIN_LOG_MODE": LogMode.get(),
+                "MODIN_LOG_FILE_SIZE": str(LogFileSize.get()),
```
I would rather you only pass environment variables if they have non-default values, as this value is shown to the user in a warning

vnlitvinov commented on modin/core/execution/ray/common/utils.py:
```diff
@@ -65,6 +68,19 @@ def initialize_ray(
     extra_init_kw = {"runtime_env": {"env_vars": {"__MODIN_AUTOIMPORT_PANDAS__": "1"}}}
+
+    # Set Modin logging envvars if not default (if set).
+    if LogMode.get() != LogMode._get_default():
```
```suggestion
    if LogMode.get_value_source() != ValueSource.DEFAULT:
```
without using protected methods (you'd have to import `ValueSource` from `modin.config` first, obviously)

for all configuration variables

YarShev commented:
Closed as outdated. Most of the changes have already been done. Remote logging is provided by such tools as ray timelime, for instance.
